### PR TITLE
Add more tests to ASSERT_STATUS_CHECKED (5) 

### PR DESCRIPTION
**Pull request by adamretter**
Fifth batch of adding more tests to ASSERT_STATUS_CHECKED.

* db_test
* db_test2

Requires https://github.com/facebook/rocksdb/pull/7819 for all tests to pass.

### CODE REVIEW

File: db/db_test2.cc [hunk 1]
```diff
@@ -849,8 +853,12 @@ TEST_F(DBTest2, WalFilterTestWithChangeBatch) {
       num_keys_added_(0) {}
     void Put(const Slice& key, const Slice& value) override {
       if (num_keys_added_ < num_keys_to_add_in_new_batch_) {
-        new_write_batch_->Put(key, value);
-        ++num_keys_added_;
+        Status s = new_write_batch_->Put(key, value);
+        if (s.ok()) {
```
Comment (mrambacher) [inline db/db_test2.cc:857]: Why not an ASSERT_OK?

File: db/db_test2.cc [hunk 2]
```diff
@@ -1806,14 +1826,15 @@ TEST_F(DBTest2, CompressionOptions) {
...
-        dbfull()->TEST_WaitForCompact();
+        ASSERT_OK(dbfull()->TEST_WaitForCompact());
       }
 
       // Make sure that we wrote enough to check all 7 levels
       ASSERT_EQ(listener->max_level_checked, 6);
 
       // Make sure database content is the same as key_value_written
       std::unique_ptr<Iterator> db_iter(db_->NewIterator(ReadOptions()));
+      ASSERT_OK(db_iter->status());
```
Comment (mrambacher) [inline db/db_test2.cc:1837]: Please move this after the for loop

File: db/db_test2.cc [hunk 3]
```diff
@@ -2388,6 +2409,7 @@ TEST_F(DBTest2, TestPerfContextIterCpuTime) {
   // CPU timing is not enabled with kEnableTimeExceptForMutex
   SetPerfLevel(PerfLevel::kEnableTimeExceptForMutex);
   Iterator* iter = db_->NewIterator(ReadOptions());
+  ASSERT_OK(iter->status());
```
Comment (mrambacher) [inline db/db_test2.cc:2412]: Move to after the last iter->Valid() check

File: db/db_test2.cc [hunk 4]
```diff
@@ -2422,6 +2444,7 @@ TEST_F(DBTest2, TestPerfContextIterCpuTime) {
 
   SetPerfLevel(PerfLevel::kEnableTimeAndCPUTimeExceptForMutex);
   iter = db_->NewIterator(ReadOptions());
+  ASSERT_OK(iter->status());
```
Comment (mrambacher) [inline db/db_test2.cc:2447]: Move to after the last iter->Valid() check

File: db/db_test2.cc [hunk 5]
```diff
@@ -2636,6 +2659,7 @@ TEST_F(DBTest2, ReadAmpBitmap) {
 
     // Make sure we read every thing in the DB (which is smaller than our cache)
     Iterator* iter = db_->NewIterator(ReadOptions());
+    ASSERT_OK(iter->status());
```
Comment (mrambacher) [inline db/db_test2.cc:2662]: Move after for loop

File: db/db_test2.cc [hunk 6]
```diff
@@ -3190,14 +3222,15 @@ TEST_F(DBTest2, IterRaceFlush1) {
...
-    Flush();
+    ASSERT_OK(Flush());
     TEST_SYNC_POINT("DBTest2::IterRaceFlush:2");
   });
 
   // iterator is created after the first Put(), so it should see either
   // "v1" or "v2".
   {
     std::unique_ptr<Iterator> it(db_->NewIterator(ReadOptions()));
+    ASSERT_OK(it->status());
```
Comment (mrambacher) [inline db/db_test2.cc:3233]: Move after it->Valid() check

File: db/db_test2.cc [hunk 7]
```diff
@@ -3248,15 +3282,16 @@ TEST_F(DBTest2, IterRefreshRaceFlush) {
...
+    ASSERT_OK(Flush());
     TEST_SYNC_POINT("DBTest2::IterRefreshRaceFlush:2");
   });
 
   // iterator is created after the first Put(), so it should see either
   // "v1" or "v2".
   {
     std::unique_ptr<Iterator> it(db_->NewIterator(ReadOptions()));
-    it->Refresh();
+    ASSERT_OK(it->status());
```
Comment (mrambacher) [inline db/db_test2.cc:3293]: Move after it->Valid

File: db/db_test2.cc [hunk 8]
```diff
@@ -3355,16 +3390,18 @@ TEST_F(DBTest2, MemtableOnlyIterator) {
 
   // Memtable-only iterator (read_tier=kMemtableTier); data not flushed yet.
   it = db_->NewIterator(ropt, handles_[1]);
+  ASSERT_OK(it->status());
```
Comment (mrambacher) [inline db/db_test2.cc:3393]: Not sure this line is necessary, but can leave it if you want.

File: db/db_test2.cc [hunk 9]
```diff
@@ -4241,6 +4284,7 @@ TEST_F(DBTest2, DISABLED_IteratorPinnedMemory) {
   // at each time.
   {
     std::unique_ptr<Iterator> iter(db_->NewIterator(ReadOptions()));
+    ASSERT_OK(iter->status());
```
Comment (mrambacher) [inline db/db_test2.cc:4287]: Move later...

File: db/db_test2.cc [hunk 10]
```diff
@@ -4482,8 +4526,9 @@ TEST_F(DBTest2, TestCompactFiles) {
   GetSstFiles(env_, dbname_, &files);
   ASSERT_EQ(files.size(), 2);
 
-  port::Thread user_thread1(
-      [&]() { db_->CompactFiles(CompactionOptions(), handle, files, 1); });
+  port::Thread user_thread1([&]() {
+    ASSERT_OK(db_->CompactFiles(CompactionOptions(), handle, files, 1));
```
Comment (mrambacher) [inline db/db_test2.cc:4530]: Another potential threading assert issue...

File: table/block_fetcher.h [hunk 11]
```diff
@@ -64,6 +64,8 @@ class BlockFetcher {
         memory_allocator_compressed_(memory_allocator_compressed),
         for_compaction_(for_compaction) {}
 
+  ~BlockFetcher() { status_.PermitUncheckedError(); }
```
Comment (mrambacher) [inline table/block_fetcher.h:67]: Do the tests pass if the PermitUncheckedError is done in the constructor?

File: trace_replay/trace_replay.cc [hunk 12]
```diff
@@ -208,8 +210,7 @@ Status Replayer::Replay() {
   WriteOptions woptions;
   ReadOptions roptions;
   Trace trace;
-  uint64_t ops = 0;
-  Iterator* single_iter = nullptr;
+  // uint64_t ops = 0;
```
Comment (mrambacher) [inline trace_replay/trace_replay.cc:213]: Why comment it out rather than remove it altogether?

**Review by mrambacher — COMMENTED** [review]
Thanks for all of the hard work in this PR.  That is quite a few places to check.

Generally, it is fine.  Questions on what to do with status in a thread and where to check the status in the iterator -- I prefer as one of the "last" things or after a Valid() check.

File: db/db_test.cc [hunk 13]
```diff
@@ -2046,13 +2065,13 @@ TEST_F(DBTest, DBOpen_Options) {
   options.create_if_missing = false;
   Status s = DB::Open(options, dbname, &db);
   ASSERT_TRUE(strstr(s.ToString().c_str(), "does not exist") != nullptr);
```
[the remark this answers is not in the document]
Comment (adamretter) [reply db/db_test.cc:2067]: Done.

File: db/db_test.cc [hunk 14]
```diff
@@ -2300,17 +2317,17 @@ TEST_F(DBTest, ReadonlyDBGetLiveManifestSize) {
 
 TEST_F(DBTest, GetLiveBlobFiles) {
   VersionSet* const versions = dbfull()->TEST_GetVersionSet();
-  assert(versions);
-  assert(versions->GetColumnFamilySet());
```
[the remark this answers is not in the document]
Comment (adamretter) [reply db/db_test.cc:2304]: Hmm... I just wanted things to be consistent. It seems strange to sometimes use `assert` and at other times the GTest ASSSERT* statements.

File: db/db_test.cc [hunk 15]
```diff
@@ -5139,16 +5180,20 @@ TEST_F(DBTest, FileCreationRandomFailure) {
         // update the latest successful put
         values[k] = value;
       }
+
       // But everything before we simulate the failure-test should succeed.
       if (j < kRandomFailureTest) {
         ASSERT_OK(s);
+      } else {
+        s.PermitUncheckedError();
```
[the remark this answers is not in the document]
Comment (adamretter) [reply db/db_test.cc:5188]: Done.

File: db/db_test.cc [hunk 16]
```diff
@@ -4431,8 +4470,10 @@ TEST_P(DBTestWithParam, PreShutdownMultipleCompaction) {
     }
 
     Status s = env_->GetThreadList(&thread_list);
-    for (auto thread : thread_list) {
-      operation_count[thread.operation_type]++;
+    if (s.ok()) {
+      for (auto thread : thread_list) {
+        operation_count[thread.operation_type]++;
+      }
```
[the remark this answers is not in the document]
Comment (adamretter) [reply db/db_test.cc:4476]: Fixed.

File: db/db_test.cc [hunk 17]
```diff
@@ -3886,9 +3919,9 @@ TEST_F(DBTest, WriteSingleThreadEntry) {
   std::vector<port::Thread> threads;
   dbfull()->TEST_LockMutex();
   auto w = dbfull()->TEST_BeginWrite();
-  threads.emplace_back([&] { Put("a", "b"); });
+  threads.emplace_back([&] { ASSERT_OK(Put("a", "b")); });
```
[the remark this answers is not in the document]
Comment (adamretter) [reply db/db_test.cc:3922]: @mrambacher I finally tracked down the earlier PR - https://github.com/facebook/rocksdb/pull/7754
I have made some changes now which are inline with the advice in that PR.

File: db/db_test.cc [hunk 18]
```diff
@@ -5848,8 +5893,8 @@ TEST_F(DBTest, CompactFilesShouldTriggerAutoCompaction) {
   SyncPoint::GetInstance()->EnableProcessing();
 
   port::Thread manual_compaction_thread([&]() {
-      auto s = db_->CompactFiles(CompactionOptions(),
-          db_->DefaultColumnFamily(), input_files, 0);
+    ASSERT_OK(db_->CompactFiles(CompactionOptions(), db_->DefaultColumnFamily(),
+                                input_files, 0));
```
[the remark this answers is not in the document]
Comment (adamretter) [reply db/db_test.cc:5897]: I have made some changes now which are inline with the advice in https://github.com/facebook/rocksdb/pull/7754

File: db/db_test2.cc (hunk 1 above)
Comment (adamretter) [reply db/db_test2.cc:857]: This is inside a callback and so I wasn't certain that it was safe to use ASSERT_* here.

File: db/db_test2.cc (hunk 2 above)
Comment (adamretter) [reply db/db_test2.cc:1837]: Done

File: db/db_test2.cc [hunk 19]
```diff
@@ -215,13 +215,14 @@ TEST_P(PrefixFullBloomWithReverseComparator,
...
 
-  dbfull()->Flush(FlushOptions());
+  ASSERT_OK(dbfull()->Flush(FlushOptions()));
 
   if (bbto.block_cache) {
     bbto.block_cache->EraseUnRefEntries();
   }
 
   std::unique_ptr<Iterator> iter(db_->NewIterator(ReadOptions()));
+  ASSERT_OK(iter->status());
```
[the remark this answers is not in the document]
Comment (adamretter) [reply db/db_test2.cc:225]: Done.

File: db/db_test2.cc (hunk 9 above)
Comment (adamretter) [reply db/db_test2.cc:4287]: Done

File: db/db_test2.cc (hunk 8 above)
Comment (adamretter) [reply db/db_test2.cc:3393]: You are right. I have removed it.

File: db/db_test2.cc (hunk 7 above)
Comment (adamretter) [reply db/db_test2.cc:3293]: Done

File: db/db_test2.cc (hunk 6 above)
Comment (adamretter) [reply db/db_test2.cc:3233]: Done

File: db/db_test2.cc (hunk 5 above)
Comment (adamretter) [reply db/db_test2.cc:2662]: Done

File: db/db_test2.cc (hunk 4 above)
Comment (adamretter) [reply db/db_test2.cc:2447]: Done

File: db/db_test2.cc (hunk 3 above)
Comment (adamretter) [reply db/db_test2.cc:2412]: Done.

File: db/db_test2.cc (hunk 10 above)
Comment (adamretter) [reply db/db_test2.cc]: I have made some changes now which are inline with the advice in https://github.com/facebook/rocksdb/pull/7754

File: trace_replay/trace_replay.cc (hunk 12 above)
Comment (adamretter) [reply trace_replay/trace_replay.cc:213]: Good point. I have now removed it.

File: table/block_fetcher.h (hunk 11 above)
Comment (adamretter) [reply table/block_fetcher.h:67]: Fixed.

File: db/db_test.cc (hunk 14 above)
[the remark this answers is not in the document]
Comment (pdillinger) [reply db/db_test.cc:2304]: These apparently useless non-nullity checks are hints for clang-analyze not to generate false reports. I don't believe clang-analyze understands GTest ASSERT*. (I always recommend "why" comments where justification is not clear.)

[the remark this answers is not in the document]
Comment (adamretter) [reply db/db_test.cc:2304]: @pdillinger So should these be changed back? There are many changes like these in lots of tests - does that also then mean that we need to change these in all tests?

[the remark this answers is not in the document]
Comment (pdillinger) [reply db/db_test.cc:2304]: Consider the CircleCI build-linux-clang10-clang-analyze to be authoritative. If it's happy, we should be good. (Hopefully our internal clang-analyze runs will agree with it; we can worry about any disagreements between the two as they happen.)

**Comment by akankshamahajan15** [comment]
@adamretter  DBTest.ThreadLocalPtrDeadlock is failing in ASAN check and ASSERT_STATUS_CHECKED.

**Comment by akankshamahajan15** [comment]
@adamretter Hi Adam, were you able to resolve the failure?

**Comment by akankshamahajan15** [comment]
@adamretter, Hi Adam. It's being a long time this PR has been in pending queue. Is there any update on this?

**Comment by adamretter** [comment]
Hi @akankshamahajan15 I rebased it this morning atop HEAD. I spent some time comparing and running the tests before and after rebase. Unfortunately after rebase (after resolving the merge conflicts) it now shows that there are several more failing tests than before, this is in part due to two things:

1. new tests have been added to db_test and db_test2 since this original PR, and those tests do not pass ASC.
2. changes have been made since this original PR, and those changes do not pass ASC (possibly because their tests were added to db_test and db_test2).

Is it best to push the rebased version and start resolving the new failures from here? Another option could be to split this PR into two, one for db_test and one for db_test2 as they are both quite large. What do you think? Let me know and I am happy to try and proceed.

**Comment by akankshamahajan15** [comment]
> Hi @akankshamahajan15 I rebased this PR this morning atop HEAD. I spent some time comparing and running the tests before and after rebase. Unfortunately after rebase (after resolving the merge conflicts) it now shows that there are several more failing tests than before, this is in part due to two things:
> 
> 1. new tests have been added to db_test and db_test2 since this original PR, and those tests do not pass ASC.
> 2. changes have been made since this original PR, and those changes do not pass ASC (possibly because their tests were added to db_test and db_test2).
> 
> Is it best to push the rebased version of this PR and start resolving the new failures from here? Another option could be to split this PR into two, one for db_test and one for db_test2 as they are both quite large. What do you think? Let me know and I am happy to try and proceed.
> 
> p.s. the original issue `DBTest.ThreadLocalPtrDeadlock` failure still stands, but I could take another stab at it also.

Thanks a lot for taking it up. I think we If we can separate the two tests, we can atleast land one test otherwise more tests will keep getting added in future. Also we have techdebt week next week so let me know if you need any help. I can take a look at DBTest.ThreadLocalPtrDeadlock next week. It would be great if we can land this one.

**Comment by mrambacher** [comment]
> > Hi @akankshamahajan15 I rebased this PR this morning atop HEAD. I spent some time comparing and running the tests before and after rebase. Unfortunately after rebase (after resolving the merge conflicts) it now shows that there are several more failing tests than before, this is in part due to two things:
> > 
> > 1. new tests have been added to db_test and db_test2 since this original PR, and those tests do not pass ASC.
> > 2. changes have been made since this original PR, and those changes do not pass ASC (possibly because their tests were added to db_test and db_test2).
> > 
> > Is it best to push the rebased version of this PR and start resolving the new failures from here? Another option could be to split this PR into two, one for db_test and one for db_test2 as they are both quite large. What do you think? Let me know and I am happy to try and proceed.
> > p.s. the original issue `DBTest.ThreadLocalPtrDeadlock` failure still stands, but I could take another stab at it also.
> 
> Thanks a lot for taking it up. I think we If we can separate the two tests, we can atleast land one test otherwise more tests will keep getting added in future. Also we have techdebt week next week so let me know if you need any help. I can take a look at DBTest.ThreadLocalPtrDeadlock next week. It would be great if we can land this one.

I am also willing to look into these next week, but would rather not re-invent the wheel.  If we can get whatever is passing into master (without the ASC-okay flag), that would be great.

**Pull request closed without merging**